### PR TITLE
TMP commit for draft PR

### LINKED ISSUE #1930: launch processes on windows
* What the feature does

Integrates the pwintools debugger to pwntools See: https://github.com/masthoon/pwintools

* Why the feature should exist

As told in this issue : https://github.com/masthoon/pwintools/issues/4 pwintools should have been written in upstream of pwntools. I want to remediate to this issue.

* What tests should be included

Some tests about processs.

I will make a pull request. Assign me to this issue please.

### PR DESCRIPTION
**Pull request by gogo2464**
This is a draft PR to show it to a friend. Please ignore it for the moment.

Closes https://github.com/Gallopsled/pwntools/issues/1930

# Pwntools Pull Request

Thanks for contributing to Pwntools!  Take a moment to look at [`CONTRIBUTING.md`][contributing] to make sure you're familiar with Pwntools development.

Please provide a high-level explanation of what this pull request is for.

## Testing

Pull Requests that introduce new code should try to add doctests for that code.  See [`TESTING.md`][testing] for more information.

## Target Branch

Depending on what the PR is for, it needs to target a different branch.

You can always [change the branch][change] after you create the PR if it's against the wrong branch.

| Branch   | Type of PR                                                       |
| -------- | ---------------------------------------------------------------- |
| `dev`    | New features, and enhancements
| `dev`    | Documentation fixes and new tests
| `stable` | Bug fixes that affect the current `stable` branch
| `beta`   | Bug fixes that affect the current `beta` branch, but not `stable`
| `dev`    | Bug fixes for code that has never been released

[contributing]: https://github.com/Gallopsled/pwntools/blob/dev/CONTRIBUTING.md
[testing]: https://github.com/Gallopsled/pwntools/blob/dev/TESTING.md
[change]: https://github.com/blog/2224-change-the-base-branch-of-a-pull-request

## Changelog

After creating your Pull Request, please add and push a commit that updates the changelog for the appropriate branch.  
You can look at the existing changelog for examples of how to do this.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -4,3 +4,4 @@
 
 # Amazing command
 -e .
+git+git://github.com/hakril/PythonForWindows
```
Comment (Arusekk) [inline requirements.txt:7]: First thing, you should rather urge hakril to make a beta release than add a 'live' version here, and second thing, move any dependency declaration to setup.py where they belong in pwntools (I would personally prefer requirements.txt too, but if they are all there, there is no valid reason to add them here).

File: pwnlib/tubes/process.py [hunk 2]
```diff
@@ -233,145 +308,175 @@ def __init__(self, argv = None,
...
+                    self.suid = st.st_uid
+                if (st.st_mode & stat.S_ISGID):
+                    self.sgid = st.st_gid
+
+        elif context.os == "windows":
+            executable_val = executable
+            self.flags = flags
+            self.stdhandles = not nostdhandles
+
+            self.debuggerpath = r'C:\Program Files (x86)\Windows Kits\10\Debuggers\x64\windbg.exe'
```
Comment (Arusekk) [inline pwnlib/tubes/process.py:460]: Pinning `C`, `x64`, `x86` and `10`, huh? Not too great, maybe use [`glob`](https://docs.python.org/dev/library/glob.html), and variables like `os.environ['PROGRAMFILES(X86)']`, if not some registry, I guess (I have not done any nt for ages, so no idea how it works now).

Is this `debuggerpath` attribute even used anywhere else, or am I missing something?

File: pwnlib/tubes/process.py [hunk 3]
```diff
@@ -608,9 +713,10 @@ def __getattr__(self, attr):
         """Permit pass-through access to the underlying process object for
         fields like ``pid`` and ``stdin``.
         """
-        if hasattr(self.proc, attr):
-            return getattr(self.proc, attr)
-        raise AttributeError("'process' object has no attribute '%s'" % attr)
+        if context.os == "linux":
```
Comment (Arusekk) [inline pwnlib/tubes/process.py:716]: Hey hey, no need to be so cautious, how is this even desired? What about `p.terminate()` or `p.returncode`, for example?

File: pwnlib/tubes/process.py [hunk 4]
```diff
@@ -734,25 +840,31 @@ def connected_raw(self, direction):
...
+            if not self._stop_noticed:
+                try:
+                    self.proc.kill()
+                    self.proc.wait()
+                    self._stop_noticed = time.time()
+                    self.info('Stopped process %r (pid %i)' % (self.program, self.pid))
+                except OSError:
+                    pass
+
+        if context.os == "windows":#TODO : close the windows process
```
Comment (Arusekk) [inline pwnlib/tubes/process.py:864]: Handles need to be closed, too; python has the concept of 'file-like' objects, that should be implemented transparently for nt pipes as well as standard pipes. Why not use the handles returned by Python API with e.g. `.fileno()`?

File: pwnlib/tubes/process.py [hunk 5]
```diff
@@ -734,25 +840,31 @@ def connected_raw(self, direction):
             return not self.proc.stdout.closed
 
     def close(self):
-        if self.proc is None:
-            return
+        if context.os == "linux":
```
Comment (Arusekk) [inline pwnlib/tubes/process.py:843]: Try to `if` out only the pieces of code that actually don't work; clearly the `if self.proc is None:` is os-agnostic, so is `self.poll`.

File: pwnlib/tubes/process.py [hunk 6]
```diff
@@ -41,7 +50,72 @@ class PTY(object): pass
...
+PeekNamedPipeParams = ((1, 'hNamedPipe'), (1, 'lpBuffer'), (1, 'nBufferSize'), (1, 'lpBytesRead'), (1, 'lpTotalBytesAvail'), (1, 'lpBytesLeftThisMessage'))
+
+@windows.winproxy.Kernel32Proxy('PeekNamedPipe', deffunc_module=sys.modules[__name__])
+def PeekNamedPipe(hNamedPipe):
+    lpTotalBytesAvail = gdef.DWORD()
+    PeekNamedPipe.ctypes_function(hNamedPipe, None, 0, None, lpTotalBytesAvail, None)
+    return lpTotalBytesAvail.value
+
+class Pipe(object):
+    """Windows pipe support"""
```
Comment (Arusekk) [inline pwnlib/tubes/process.py:73]: I thought [`os.pipe`](https://docs.python.org/dev/library/os.html#os.pipe) was capable of creating windows anonymous pipes as well.

**Review by Arusekk — CHANGES_REQUESTED** [review]
Thanks for submitting the PR, I see that it is not yet marked ready for review, but I thought I should point some things out as early as possible to save your time, especially that the PR has been here for a while now, and it looks like you would like to integrate some of it into pwntools, for real. Marking as Changes Requested only to mark the PR as not ready (in the current shape).

I see that there are still some verbose debugs scattered throughout the core modules; try to put these into separate temporary scripts, like `winwin.py` or `todo123.py`.

As for the unconditional `import windows`, they are all too much, and break normal standards-compliant oses, and should be put preferably into a separate tube, or imported from a separate module on a use-site basis. In files like `pwnlib/tubes/winprocess.py` or `pwnlib/util/ntcompat.py`. They should implement nt-specific additional features and workarounds for the missing functionalities of nt, which all other oses already have, or if some feature is completely missing from nt, and implemented upside down, they should be made into a new class. Of course if you have a better idea, it will be OK, as long as it cannot be called unelegant.

**Comment by gogo2464** [comment]
All your comments are true. My PR is currently just a copy and paste of https://github.com/masthoon/pwintools/blob/master/pwintools.py but at this moment my proirity is to make it just work. pwintools does not works on python3. Then I even need a PR for pwintools and even for PythonForWindows which works only in python3 see: https://github.com/hakril/PythonForWindows/issues/23.

When I will have fixed all that I will check out your comments.

**Comment by Arusekk** [comment]
Closing as this is almost 100% not-ready, and there has been no positive feedback happening in this PR.

**Pull request closed without merging**